lib/libdl/dlfcn.c

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019 Linaro limited
+ */
+
+#include <ctype.h>
+#include <dlfcn.h>
+#include <pta_system.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee_api.h>
+#include <tee_internal_api_extensions.h>
+
+static TEE_TASessionHandle sess = TEE_HANDLE_NULL;
+
+static TEE_Result invoke_system_pta(uint32_t cmd_id, uint32_t param_types,
+				    TEE_Param params[TEE_NUM_PARAMS])
+{
+	const TEE_UUID core_uuid = PTA_SYSTEM_UUID;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (!sess) {
+		res = TEE_OpenTASession(&core_uuid, 0, 0, NULL, &sess, NULL);
+		if (res)
+			return res;
+	}
+	return TEE_InvokeTACommand(sess, 0, cmd_id, param_types, params, NULL);
+}
+
+static void *param_to_ptr(TEE_Param *param)
+{
+	vaddr_t va = param->value.a;
+
+#ifdef ARM64
+	va |= ((vaddr_t)param->value.b) << 32;
+#endif
+	return (void *)va;
+}
+
+struct dl_handle {
+	TEE_UUID uuid;
+};
+
+void *dlopen(const char *filename, int flags)
+{
+	TEE_Param params[TEE_NUM_PARAMS] = { };
+	struct dl_handle *h = NULL;
+	uint32_t param_types = 0;
+	void *ret = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_UUID uuid = { };
+
+	res = tee_parse_uuid(filename, &uuid);
+	if (res)
+		goto err;
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				      TEE_PARAM_TYPE_VALUE_INPUT,
+				      TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE);
+
+	params[0].memref.buffer = (void *)&uuid;
+	params[0].memref.size = sizeof(uuid);
+	params[1].value.a = flags;
+
+	res = invoke_system_pta(PTA_SYSTEM_DLOPEN, param_types, params);
+	if (res)
+		goto err;
+	h = malloc(sizeof(*h));
+	if (!h)
+		goto err;
+	h->uuid = uuid;
+	ret = (void *)h;
+err:
+	return ret;
+}
+
+int dlclose(void *handle)
+{
+	free(handle);
+	return 0;
+}
+
+void *dlsym(void *handle, const char *symbol)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_Param params[TEE_NUM_PARAMS] = { };
+	struct dl_handle *h = handle;
+	uint32_t param_types = 0;
+	TEE_UUID zero = { };
+	void *ptr = NULL;
+
+	if (!symbol)
+		return NULL;
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				      TEE_PARAM_TYPE_MEMREF_INPUT,
+				      TEE_PARAM_TYPE_VALUE_OUTPUT,
+				      TEE_PARAM_TYPE_NONE);
+
+	if (h) {
+		params[0].memref.buffer = &h->uuid;
+		params[0].memref.size = sizeof(h->uuid);
+	} else {
+		/*
+		 * Currently we can't pass a NULL buffer pointer to a TA, let's
```

Comment:
I guess you can drop "Currently" because I don't think this will change once it's there.